url64d634b8f0_agdturner_generic_tgz-pJ8-uk_ac_leeds_ccg_andyt_generic_visualisation_charts_Generic_BarChartJ8/src/uk/ac/leeds/ccg/andyt/generic/io/Generic_StaticIO.java

```diff
@@ -0,0 +1,2156 @@
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.PrintWriter;
+import java.io.StreamTokenizer;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import uk.ac.leeds.ccg.andyt.generic.core.Generic_ErrorAndExceptionHandler;
+//import java.nio.file.S
+
+//TODO http://java.sun.com/docs/books/tutorial/essential/io/legacy.html#mapping
+// http://java.sun.com/docs/books/tutorial/essential/io/fileio.html
+/**
+ * Class of static methods for helping with reading and writing (primarily)
+ * to/from file.
+ */
+public class Generic_StaticIO {
+
+    public static final String CLASSNAME = "Generic_StaticIO";
+
+    /**
+     * Creates a new instance of ObjectIO
+     */
+    public Generic_StaticIO() {
+    }
+
+    public static TreeSet<String> recursiveFileList(File file) {
+        TreeSet<String> result = new TreeSet<String>();
+        if (file.isDirectory()) {
+            File[] files = file.listFiles();
+            for (File file1 : files) {
+                TreeSet<String> subresult = recursiveFileList(file1);
+                result.addAll(subresult);
+            }
+            return result;
+        } else {
+            result.add(file.toString());
+            return result;
+        }
+    }
+
+    public static TreeSet<String> recursiveFileList(File file, int depth) {
+        TreeSet<String> result = new TreeSet<String>();
+        if (depth != 0) {
+            if (file.isDirectory()) {
+                File[] files = file.listFiles();
+                for (File f : files) {
+                    TreeSet<String> subresult = recursiveFileList(f, depth - 1);
+                    result.addAll(subresult);
+                }
+                return result;
+            } else {
+                result.add(file.toString());
+                return result;
+            }
+        }
+        result.add(file.toString());
+        return result;
+    }
+
+    /**
+     * Write object to file
+     *
+     * @param object
+     * @param f
+     */
+    public static void writeObject(
+            Object object,
+            File f) {
+        try {
+            ObjectOutputStream oos;
+            oos = getObjectOutputStream(f);
+            oos.writeObject(object);
+            oos.flush();
+            oos.close();
+        } catch (IOException e) {
+            System.err.print(e.getMessage());
+            e.printStackTrace(System.err);
+            System.exit(Generic_ErrorAndExceptionHandler.IOException);
+        }
+    }
+
+    /**
+     * Read Object from File
+     *
+     * @param f
+     * @return
+     */
+    public static Object readObject(
+            File f) {
+        Object result = null;
+        if (f.length() != 0) {
+            try {
+                ObjectInputStream ois;
+                ois = getObjectInputStream(f);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *ois*. The resource is referenced at line 128. The resource is closed at line 129. There are other execution paths that do not contain closure statements, for example, when ObjectInputStream.readObject throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).